src/redisearch_rs/result_processor/src/test_utils.rs

```diff
@@ -0,0 +1,210 @@
+use crate::{Context, Error, ResultProcessor, ResultProcessorWrapper};
+use std::{
+    pin::Pin,
+    ptr::{self, NonNull},
+};
+
+/// Create a ResultProcessor from an `Iterator` for testing purposes
+pub fn from_iter<I>(i: I) -> IterResultProcessor<I::IntoIter>
+where
+    I: IntoIterator<Item = ffi::SearchResult>,
+{
+    IterResultProcessor {
+        iter: i.into_iter(),
+    }
+}
+
+/// ResultProcessor that yields items from an inner `Iterator`
+#[derive(Debug)]
+pub struct IterResultProcessor<I> {
+    iter: I,
+}
+
+impl<I> ResultProcessor for IterResultProcessor<I>
+where
+    I: Iterator<Item = ffi::SearchResult>,
+{
+    const TYPE: ffi::ResultProcessorType = ffi::ResultProcessorType_RP_MAX + 1;
+
+    fn next(&mut self, _cx: Context, out: &mut ffi::SearchResult) -> Result<Option<()>, Error> {
+        if let Some(res) = self.iter.next() {
+            *out = res;
+            Ok(Some(()))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+/// A result processor that returns the provided result.
+pub struct ResultRP {
+    res: Option<Result<Option<()>, Error>>,
+}
+impl ResultRP {
+    pub fn new_err(error: Error) -> Self {
+        Self {
+            res: Some(Err(error)),
+        }
+    }
+    pub fn new_ok_some() -> Self {
+        Self {
+            res: Some(Ok(Some(()))),
+        }
+    }
+    pub fn new_ok_none() -> Self {
+        Self {
+            res: Some(Ok(None)),
+        }
+    }
+}
+impl ResultProcessor for ResultRP {
+    const TYPE: ffi::ResultProcessorType = ffi::ResultProcessorType_RP_MAX;
+
+    fn next(&mut self, _cx: Context, _res: &mut ffi::SearchResult) -> Result<Option<()>, Error> {
+        self.res.take().unwrap()
+    }
+}
+
+/// Return the default value for [`ffi::SearchResult`]
+// FIXME replace with Default::default once `ffi::SearchResult` is ported to Rust
+pub const fn default_search_result() -> ffi::SearchResult {
+    const SEARCH_RESULT_INIT: ffi::SearchResult = ffi::SearchResult {
+        docId: 0,
+        score: 0.0,
+        scoreExplain: ptr::null_mut(),
+        dmd: ptr::null(),
+        indexResult: ptr::null_mut(),
+        rowdata: ffi::RLookupRow {
+            sv: ptr::null(),
+            dyn_: ptr::null_mut(),
+            ndyn: 0,
+        },
+        flags: 0,
+    };
+    SEARCH_RESULT_INIT
+}
+
+/// A mock implementation of the "result processor chain" part of the `QueryIterator`
+///
+/// It acts as an owning collection of linked result processors.
+pub struct Chain {
+    result_processors: Vec<NonNull<crate::Header>>,
+}
+
+impl Chain {
+    pub const fn new() -> Self {
+        Self {
+            result_processors: Vec::new(),
+        }
+    }
+
+    /// Append a new result processor at the end of the chain. It will have its `upstream`
+    /// field set to the previous last result processor.
+    pub fn append<P>(&mut self, mut result_processor: Pin<Box<ResultProcessorWrapper<P>>>)
```

Comment:
I think it'd nicer to take `ResultProcessorWrapper<P>` as input here, moving the box-pinning inside `append`.

Reply:
changed 👍🏻